### PR TITLE
Use stable identifiers for faster joins

### PR DESCRIPTION
**Pull request by DMRobertson**
No attempt to maintain backwards compatibility here.

Synapse issue: https://github.com/matrix-org/synapse/issues/14828
Synapse PR: https://github.com/matrix-org/synapse/pull/14832
Complement PR: https://github.com/matrix-org/complement/pull/583
GMSL PR: https://github.com/matrix-org/gomatrixserverlib/pull/350
MSC: https://github.com/matrix-org/matrix-spec-proposals/pull/3706
Spec diff: https://github.com/matrix-org/matrix-spec/pull/1393

TODO: once GMSL gets merged, update go.mod to point at main branch of GMSL

### CODE REVIEW

**Comment by DMRobertson** [comment]
Ah. Looks like I need to update to changes in GMSL?

File: tests/federation_room_join_test.go [hunk 1]
```diff
@@ -481,26 +483,26 @@ func TestSendJoinPartialStateResponse(t *testing.T) {
...
-	sendJoinResp, err := fedClient.SendJoinPartialState(context.Background(), "hs1", joinEvent)
+	sendJoinResp, err := fedClient.SendJoinPartialState(context.Background(), origin, "hs1", joinEvent)
 	if err != nil {
 		t.Fatalf("send_join failed: %v", err)
 	}
 
-	if !sendJoinResp.PartialState {
 		t.Skip("Server does not support partial_state")
+	if !sendJoinResp.MembersOmitted {
 	}
```
Comment (squahtx) [inline tests/federation_room_join_test.go:506]: Not sure what's going on here?

**Review by squahtx — APPROVED** [review]
lgtm.
Thanks for dealing with the gomatrixserverlib update. It looks like a right pain.

File: tests/federation_room_join_test.go (hunk 1 above)
Comment (DMRobertson) [reply tests/federation_room_join_test.go:506]: I think this is poor git-hygiene, will fix

**Pull request merged**